jupyterhub/scopes.py

```diff
@@ -13,6 +13,42 @@ class Scope(Enum):
     ALL = True
 
 
+def _intersect_scopes(token_scopes, owner_scopes):
+    """Compares the permissions of token and its owner including horizontal filters
+    Returns the intersection of the two sets of scopes"""
+    owner_parsed_scopes = parse_scopes(owner_scopes)
+    token_parsed_scopes = parse_scopes(token_scopes)
+
+    common_bases = owner_parsed_scopes.keys() & token_parsed_scopes.keys()
+
+    common_filters = {}
+    for base in common_bases:
+        if owner_parsed_scopes[base] == Scope.ALL:
+            common_filters[base] = token_parsed_scopes[base]
+        elif token_parsed_scopes[base] == Scope.ALL:
+            common_filters[base] = owner_parsed_scopes[base]
+        else:
+            common_entities = (
+                owner_parsed_scopes[base].keys() & token_parsed_scopes[base].keys()
+            )
+            common_filters[base] = {
```

Comment:
From in-person review, add notes here about what's not handled yet:

- filters intersect correctly with ALL and exact filter matches (i.e. `users!user=x & read:users:name` -> `read:users:name!user=x`)
- filters *do not* intersect with containing filters, i.e. `users!group=x` does not intersect with `users!user=y` even if user `y` is in group `x`. Same for user->server.

We should probably log a warning when this happens (group, user filters both present, result may be lower than intended permissions).

Reply:
Note and warning added.